### ISSUE TITLE
dm push sometimes hangs on 'awaiting transfer...'

### ISSUE THREAD

**Issue by prisamuel** [dotmesh-io/dotmesh#336]
We're seeing this a lot when running the back up script.

Multi node cluster tests also hang on awaiting transfer.
http://gitlab.dotmesh.io:9999/dotmesh/dotmesh-sync/-/jobs/53061

**Comment by binocarlos** [comment]
A log of a script we were running that shows the output for this problem:

```
h 'chanezon/moby_counter'
dm push backuphub 'chanezon/moby_counter' 'branch_a' --remote-name 'chanezon/moby_counter'
Pushing chanezon/moby_counter to backuphub:chanezon/moby_counter
Calculating...
Awaiting transfer...
dm push backuphub 'chanezon/moby_counter' 'master' --remote-name 'chanezon/moby_counter'
Pushing chanezon/moby_counter to backuphub:chanezon/moby_counter
Calculating...
Awaiting transfer...
dm switch 'colinb/moby_counter'
dm push backuphub 'colinb/moby_counter' 'branch_a' --remote-name 'colinb/moby_counter'
Pushing colinb/moby_counter to backuphub:colinb/moby_counter
Calculating...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
^Cdm push backuphub 'colinb/moby_counter' 'master' --remote-name 'colinb/moby_counter'
Pushing colinb/moby_counter to backuphub:colinb/moby_counter
Calculating...
Awaiting transfer...
dm switch 'dan/wordpressdotmesh_wordpress'
dm push backuphub 'dan/wordpressdotmesh_wordpress' 'master' --remote-name 'dan/wordpressdotmesh_wordpress'
Pushing dan/wordpressdotmesh_wordpress to backuphub:dan/wordpressdotmesh_wordpress
Calculating...
Awaiting transfer...
dm switch 'dan/wordpressdotmesh_wp'
dm push backuphub 'dan/wordpressdotmesh_wp' 'master' --remote-name 'dan/wordpressdotmesh_wp'
Pushing dan/wordpressdotmesh_wp to backuphub:dan/wordpressdotmesh_wp
Calculating...
Awaiting transfer...
dm switch 'ericlugo/moby_counter'
dm push backuphub 'ericlugo/moby_counter' 'branch_a' --remote-name 'ericlugo/moby_counter'
Pushing ericlugo/moby_counter to backuphub:ericlugo/moby_counter
Calculating...
finished 9.50 KB / 9.50 KB [==========================] 100.00% 0.08 MiB/s (2/2)
Done!
dm push backuphub 'ericlugo/moby_counter' 'master' --remote-name 'ericlugo/moby_counter'
Pushing ericlugo/moby_counter to backuphub:ericlugo/moby_counter
Calculating...
Awaiting transfer...
dm switch 'eveld/moby_counter'
dm push backuphub 'eveld/moby_counter' 'branch_a' --remote-name 'eveld/moby_counter'
Pushing eveld/moby_counter to backuphub:eveld/moby_counter
Calculating...
Awaiting transfer...
dm push backuphub 'eveld/moby_counter' 'master' --remote-name 'eveld/moby_counter'
Pushing eveld/moby_counter to backuphub:eveld/moby_counter
Calculating...
Awaiting transfer...
dm switch 'f2hex/moby_counter'
dm push backuphub 'f2hex/moby_counter' 'branch_a' --remote-name 'f2hex/moby_counter'
Pushing f2hex/moby_counter to backuphub:f2hex/moby_counter
Calculating...
Awaiting transfer...
dm push backuphub 'f2hex/moby_counter' 'master' --remote-name 'f2hex/moby_counter'
Pushing f2hex/moby_counter to backuphub:f2hex/moby_counter
Calculating...
Awaiting transfer...
dm switch 'fintanr/moby_counter'
dm push backuphub 'fintanr/moby_counter' 'branch_a' --remote-name 'fintanr/moby_counter'
Pushing fintanr/moby_counter to backuphub:fintanr/moby_counter
Calculating...
Awaiting transfer...
dm push backuphub 'fintanr/moby_counter' 'master' --remote-name 'fintanr/moby_counter'
Pushing fintanr/moby_counter to backuphub:fintanr/moby_counter
Calculating...
Awaiting transfer...
dm switch 'gokmen/moby_counter'
dm push backuphub 'gokmen/moby_counter' 'branch_a' --remote-name 'gokmen/moby_counter'
Pushing gokmen/moby_counter to backuphub:gokmen/moby_counter
Calculating...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
^Cdm push backuphub 'gokmen/moby_counter' 'master' --remote-name 'gokmen/moby_counter'
Pushing gokmen/moby_counter to backuphub:gokmen/moby_counter
Calculating...
Awaiting transfer...
dm switch 'hleclerc/moby_counter'
dm push backuphub 'hleclerc/moby_counter' 'branch_a' --remote-name 'hleclerc/moby_counter'
Pushing hleclerc/moby_counter to backuphub:hleclerc/moby_counter
Calculating...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
Awaiting transfer...
```

**Comment by binocarlos** [comment]
^ I did a ctrl+c for each `Awaiting transfer...` to progress to the next one in the list

**Comment by prisamuel** [comment]
Fixed with:
- Skipping branches with no commits on them
- Script issues, i.e stripping spaces from branch names, and getting the right branch name before attempting to do a `dm push`

**Comment by prisamuel** [comment]
Seeing this emerge on zfs-kernel-module backups
```
dm push backup-cluster-1 'zfs-kernel-modules' 'master' --remote-name 'zfs-kernel-modules'
Pushing admin/zfs-kernel-modules to backup-cluster-1:admin/zfs-kernel-modules
Calculating...
Awaiting transfer... 
Awaiting transfer... 
Awaiting transfer... 
Awaiting transfer... 
Awaiting transfer... 
Awaiting transfer... 
Awaiting transfer... 
Awaiting transfer... 
Awaiting transfer... 
Awaiting transfer... 
```

**Comment by prisamuel** [comment]
This could be related to https://github.com/dotmesh-io/dotmesh/issues/352

**Comment by lukemarsden** [comment]
This affected me in a demo, although I was on 0.3.1.

Released 0.3.2 with https://github.com/dotmesh-io/dotmesh/issues/339, but we should:
(a) inspect the logs for those clusters (ask me for access)
(b) try to reproduce the issue by re-running what I did in the demo like 10 times

**Comment by prisamuel** [comment]
It works!
